src/convolution/mod.rs

```diff
@@ -0,0 +1,109 @@
+use std::ops::Index;
+use std::f64;
+
+#[derive(Debug)]
+pub enum ConvolutionError {
+    ConvolutionIsNotSquare,
+    ConvolutionDoesNotContainNormalNumbers,
+}
+
+fn is_correctly_dimensionsed(convolution: &[f64]) -> bool {
+    let sqrt = (convolution.len() as f64).sqrt();
+    sqrt.floor() * sqrt.floor() == convolution.len() as f64 && sqrt.floor() as u64 % 2 == 1
+}
+
+pub struct Convolution<'a> {
+    convolution: &'a [f64],
+    size: usize,
+    min_value: f64,
+    max_value: f64,
+}
+
+impl<'a> Convolution<'a> {
+    pub fn new(values: &'a [f64]) -> Result<Convolution<'a>, ConvolutionError> {
+        if values.iter().cloned().any(
+            |x| x.is_infinite() || x.is_nan(),
+        )
+        {
+            return Err(ConvolutionError::ConvolutionDoesNotContainNormalNumbers);
```

Comment:
This error seems poorly named, since the result could absolutely contain subnormal numbers